File: python/src/intranet_sdk/intranet.py
"""Main client for the Intranet SDK."""

from typing import Optional

from .client import Config, HTTPClient
from .services import ConnectorService, UserService


class Client:
    """Client for the MiniEye Intranet API.

    This is the main entry point for interacting with the API. It provides
    access to various services through its properties.

    Attributes:
        user: User service for user-related operations
        connector: Connector service for messaging operations

    Example:
        >>> client = Client(
        ...     access_key_id="your_key_id",
        ...     access_key_secret="your_key_secret"
        ... )
        >>> user_info = client.user.get_user_info()
        >>> print(user_info.username)
    """

    def __init__(
        self,
        base_url: str = "https://intranet.minieye.tech/sys-api",
        api_key: Optional[str] = None,
        access_key_id: Optional[str] = None,
        access_key_secret: Optional[str] = None,
        user_agent: str = "minieye-intranet-sdk-python/0.1.0",
        timeout: int = 30,
    ):
        """Initialize the Intranet API client.

        Args:
            base_url: Base URL for API requests. Default: https://intranet.minieye.tech/sys-api
            api_key: API key for authentication (optional)
            access_key_id: Access Key ID for STS authentication (optional)
            access_key_secret: Access Key Secret for STS authentication (optional)
            user_agent: User agent string. Default: minieye-intranet-sdk-python/0.1.0
            timeout: Request timeout in seconds. Default: 30

        Raises:
            ValueError: If configuration is invalid
        """
        # Create configuration
        config = Config(
            base_url=base_url,
            api_key=api_key,
            access_key_id=access_key_id,
            access_key_secret=access_key_secret,
            user_agent=user_agent,
            timeout=timeout,
        )

        # Create HTTP client
        self._http_client = HTTPClient(config)

        # Initialize services
        self.user = UserService(self._http_client)
        self.connector = ConnectorService(self._http_client)

File: intranet.go
// Package intranet provides a client for accessing the MiniEye Intranet RESTful API.
package intranet

import (
	"github.com/hujia-team/intranet-sdk/client"
	"github.com/hujia-team/intranet-sdk/services"
)

// Client represents a client for the MiniEye Intranet API.
type Client struct {
	client    *client.HTTPClient
	User      services.UserService
	Connector services.ConnectorService
	ApiKey    services.ApiKeyService
}

// NewClient creates a new MiniEye Intranet API client with the given options.
func NewClient(options ...Option) (*Client, error) {
	cfg := defaultConfig()
	for _, opt := range options {
		opt(cfg)
	}

	httpClient, err := client.NewHTTPClient(cfg)
	if err != nil {
		return nil, err
	}

	// Create user service
	userService := services.NewUserService(httpClient)

	// Create connector service
	connectorService := services.NewConnectorService(httpClient)

	// Create apikey service
	apiKeyService := services.NewApiKeyService(httpClient)

	return &Client{
		client:    httpClient,
		User:      userService,
		Connector: connectorService,
		ApiKey:    apiKeyService,
	}, nil
}

// Option is a function for configuring the client.
type Option func(*client.Config)

// Config is an alias to client.Config for backward compatibility
type Config client.Config

// defaultConfig returns a default configuration.
func defaultConfig() *client.Config {
	return &client.Config{
		BaseURL:   "https://intranet.minieye.tech/sys-api",
		UserAgent: "minieye-intranet-sdk/1.0",
	}
}

// WithBaseURL sets the base URL for API requests.
func WithBaseURL(url string) Option {
	return func(c *client.Config) {
		c.BaseURL = url
	}
}

// WithAPIKey sets the API key for authentication.
func WithAPIKey(apiKey string) Option {
	return func(c *client.Config) {
		c.APIKey = apiKey
	}
}

// WithUserAgent sets the user agent for API requests.
func WithUserAgent(userAgent string) Option {
	return func(c *client.Config) {
		c.UserAgent = userAgent
	}
}

// WithAccessKeyID sets the access key ID for STS authentication.
func WithAccessKeyID(accessKeyID string) Option {
	return func(c *client.Config) {
		c.AccessKeyID = accessKeyID
	}
}

// WithAccessKeySecret sets the access key secret for STS authentication.
func WithAccessKeySecret(accessKeySecret string) Option {
	return func(c *client.Config) {
		c.AccessKeySecret = accessKeySecret
	}
}
